core/src/main/java/jenkins/util/JSONSignatureValidator.java

```diff
@@ -73,7 +75,17 @@ public FormValidation verifySignature(JSONObject o) throws IOException {
                     certs.add(c);
                 }
 
-                CertificateUtil.validatePath(certs, loadTrustAnchors(cf));
+                try {
+                    CertificateUtil.validatePath(certs, loadTrustAnchors(cf));
+                } catch (GeneralSecurityException | IOException e) {
+                    Iterable<Object> certSubjectDNs = Iterables.transform(certs, new Function<X509Certificate, Object>() {
+                        @Override
+                        public Object apply(X509Certificate cert) {
+                            return cert == null ? null : cert.getSubjectDN();
+                        }
+                    });
+                    return FormValidation.warning(e,String.format("Certificate validation failed with certs %s in %s", certSubjectDNs.toString(), name));
```

Comment:
I really liked the detailed message (see below and [JENKINS-26688](https://issues.jenkins-ci.org/browse/JENKINS-26688)). I feel it is particularly valuable when different tools are signed by different certificates (CloudBees uses its own certificate to sign its specific Downloadable Tool Installers).

```
Certificate validation failed with certs [CN=Community Update Center, O=Jenkins Project, ST=California, C=US] in downloadable hudson.tools.JDKInstaller java.security.cert.CertPathValidatorException: timestamp check failed
	at sun.security.provider.certpath.PKIXMasterCertPathValidator.validate(PKIXMasterCertPathValidator.java:129)
	...
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.security.cert.CertificateExpiredException: NotAfter: Tue Jan 27 22:04:07 CET 2015
	at sun.security.provider.certpath.PKIXMasterCertPathValidator.validate(PKIXMasterCertPathValidator.java:119)
	... 
```